mantidproject/mantid
# Label Tool adding Log Values to a plot improved

## Description of work.

The [LabelTool](https://github.com/mantidproject/mantid/blob/master/MantidPlot/src/LabelTool.cpp) currently in Mantid uses a context menu to show the information for each available log, and depending on the information in the log can hide the whole screen.
![1](https://cloud.githubusercontent.com/assets/9135965/16912836/98ed3a18-4cde-11e6-82be-7bf8c2d9653f.png)

This is now done via a Dialog, very similar to the _Sample Logs_ dialog when right-clicking a Workspace.
![2 5](https://cloud.githubusercontent.com/assets/9135965/16912875/ded0c1da-4cde-11e6-81a6-ba0ef1bfe448.png)

When left clicking on a plot using the **Label Tool**, the _Log Values_ button now shows the same context menu as _Workspaces_
![6_with_multiple_ws](https://cloud.githubusercontent.com/assets/9135965/17052701/82da920a-4ff6-11e6-8ec8-f94063166655.png)

and when the button is clicked the Sample Log Dialog appears **for the selected workspace**.
It contains the new functionality described in the issue, a specific log statistic can be selected. If the log has no statistics, instead the Value and Units columns will be displayed
![4](https://cloud.githubusercontent.com/assets/9135965/16913002/a6724ef2-4cdf-11e6-939d-763ea92e13f4.png)
![5](https://cloud.githubusercontent.com/assets/9135965/16913001/a6725582-4cdf-11e6-9e5b-293d63dfa666.png)

UML of the current structure of the classes (without methods):
![uml](https://cloud.githubusercontent.com/assets/9135965/17053274/2c82b786-4ff9-11e6-973a-a9d043df4bb5.png)

SampleLogDialogBase provides common functionality used by the child classes. The [void init(void)](https://github.com/mantidproject/mantid/commit/c3d3d2db1d79a99b43b7142e4eae7227d03dd4e1#diff-c0430e482e94c7ce3c7ee188231c87b0R86) is not virtual because the functionality should be common to child classes, it initialises the information into the QTreeWidget and loads the Log Statistics.
Derived classes have to implement their own [**ImportItem(QTreeWidgetItem)**](https://github.com/mantidproject/mantid/blob/6e4daf37057a177e9eef1d647bc4ffe6d0730bf1/MantidPlot/src/Mantid/SampleLogDialogBase.h#L78). 
## To test:


1. Functionality of the old Sample Dialog from the Workspaces Right Click Context Menu (picture 2), as the class was refactored. The functionality should **not** be changed.
   - How to test this: 
   - Use the Sample Dialog on any workspace, try importing different logs and see if they look correct/the same as the one from the Mantid Master branch or latest release version
   - Try importing multiple logs, with different filters, and see if they are correct
   - Try importing by using the button, double click, right click -> import, or any other ways that I haven't discovered
2. Functionality of the new Log Values Sample Dialog can be reached by plotting a spectrum of any workspace that has logs
   - How/What to test:
   - Use the Sample Dialog to import different logs' statistics, logs without statistics should have their value and units (if available) imported
   - Try importing multiple logs, you should not be allowed to select more than one, and it is intended
   - Try importing by using the button, double click, right click -> import, etc
   - Try importing a log from a plot that has 2 or more workspaces in it, Log Values will display all different workspaces in the plot and you can select the one you want
   - Make sure the logs displayed are _correct_ for the selected workspace

Fixes #13624 



---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
